### ISSUE TITLE
A proposal of `SobolSampler`

### ISSUE THREAD

**Issue by kstoneriv3** [optuna/optuna#1797]
## Motivation
The use of quasi-random low-discrepancy sequences in place of random sampling is known to perform better [1]. Sobol sequence is one of such sequences and is reported to perform well in [1]. 

## Description
I implemented `SobolSampler` for optuna using a package [`sobol_seq`](https://github.com/naught101/sobol_seq), which is 1) lightweight and 2) distributed under MIT license. My implementation can be checked at [my repository](https://github.com/kstoneriv3/optuna/tree/sobol-sampler/optuna/samplers/_sobol).

## Benchmark
I compared by my implementation of `SobolSampler` with existing samplers of Optuna using [kurobako](https://github.com/sile/kurobako). From 100 benchmarks per each sampler and each dataset, `SobolSampler` was reported (statistically significantly) better than `RandomSampler` for all datasets.

![hpo-bench-naval-4b38e028c25f65733b05b737c70a0401664a4bab047ab08e1c5e2395ed1e9705](https://user-images.githubusercontent.com/30857855/92428717-dbb54800-f1ca-11ea-989e-5c5a4c67135c.png)
![hpo-bench-parkinson-996baeee92a729fabb24c4d9b498bdb1477f2f88f9f85552a0b7981d7501c078](https://user-images.githubusercontent.com/30857855/92428719-dc4dde80-f1ca-11ea-9ddb-a8be9947a470.png)
![hpo-bench-protein-cbf8512fed27dedfec67a7e80c0921d6ef737a1e902569a7bac60b82e47ceab2](https://user-images.githubusercontent.com/30857855/92428721-dce67500-f1ca-11ea-921a-393cae47ec30.png)
![hpo-bench-slice-63956311448c8b8fafcdef5dcba69c838349546715298e00e22cdb8c7a48c655](https://user-images.githubusercontent.com/30857855/92428723-dd7f0b80-f1ca-11ea-969a-7c334b263ca9.png)

## TODOs (updated)
- [x] Discuss how we should support the Sobol sequence generator in Optuna
- [x] Create PR for SobolSampler (and maybe for Sobol sequence generator)
- [ ] Benchmark the CMA-ES combined with the Sobol sequences

## References
[1] [Random Search for Hyper-Parameter Optimization
](https://www.jmlr.org/papers/v13/bergstra12a.html)

**Comment by HideakiImamura** [comment]
Thanks for the interesting proposal! Implementing the `SobolSampler` and introducing it to Optuna is important to improve the sampling performance. I'm looking forward to it!

One concern is that the `SobolSampler` may only be used when the search space is static. Allowing the dynamic search space is one of the most important Optuna features. I think it's important to know how to deal with this issue.

**Comment by kstoneriv3** [comment]
Thank you for your interest! At the moment, `SobolSampler` samples from the same search space as the first trial or the one specified by the user (using sample_relative). If a parameter to sample is not included in the above space, it will just use independent sampling. So users need to make sure to specify search spaces which they want to efficiently cover.

Though I am not an expert of the Sobol sequence, I believe that Dynamic search space should be out of the scope for now because the Sobol sequence is designed to fill the unit hypercube uniformly (or with low discrepancy). Use of dynamic search space immediately means that the user is not sampling from a hypercube so the sampler might not be as efficient as one might expect even if we support dynamic search space with some ad-hoc methods. I have heard of researches on low-discrepancy sequences on the various space (e.g. a sphere surface) but they look too complicated.

**Comment by kstoneriv3** [comment]
Another proposal; using a randomized [Sobol sequence in the CMA-ES](https://dl.acm.org/doi/10.1007/978-3-319-31471-6_5) seems promising. Any opinion on this?

**Comment by nmasahiro** [comment]
> Another proposal; using a randomized [Sobol sequence in the CMA-ES](https://dl.acm.org/doi/10.1007/978-3-319-31471-6_5) seems promising. Sobol sequence can be used to sample from a multivariate normal distribution. Any opinion/comment on this?

This result seems to be promising.
One thing I'm curious about is that the population size in the paper is set to a relatively large value, 20 * dim (the recommended value of the population size of CMA-ES is 4 + floor(3 * log(dim))).
Based on the use case of Optuna where the evaluation budget is usually limited, we might want to investigate the behavior of the Sobol sampler with a small population size.

**Comment by kstoneriv3** [comment]
I agree that we need benchmark results to see if using sobol with CMA-ES works well in a realistic setting. At the moment, I am optimistic about it because the `SobolSampler` is performing better than `RandomSampler` where the budget is larger than 20 in the benchmark I presented above.

**Comment by toshihikoyanase** [comment]
Thank you for your proposal. `SobolSampler` seems promising.

> I implemented SobolSampler for optuna using a package `sobol_seq`, which is 1) lightweight and 2) distributed under MIT license. My implementation can be checked at [my repository](https://github.com/kstoneriv3/optuna/tree/sobol-sampler/optuna/samplers/_sobol).

I found that your current implementation includes a [sobol sequence generator](https://github.com/kstoneriv3/optuna/blob/sobol-sampler/optuna/samplers/_sobol/sobol_seq/sobol_seq.py) copied from http://people.sc.fsu.edu/~jburkardt/py_src/sobol/sobol.html instead of using `sobol_seq` package. 

IMO, the copied files increase the maintenance cost of the code in terms of both the bug fix and code license. How about providing the `SobolSampler` as an integration module? If it gets many users, we can re-create it as a built-in sampler. We employed a similar approach when we provide `CmaEsSampler`.

**Comment by HideakiImamura** [comment]
@toshihikoyanase Thanks for the proposal. I agree with your opinion, that is, the `SobolSampler` should be implemented as an integration module. In addition, we can implement the `SobolSampler` in `samplers` if we implement the algorithms of the `sobol_sq` package in the future.

@kstoneriv3 Do you have any concerns? I would be very happy to have your contribution.

**Comment by kstoneriv3** [comment]
@HideakiImamura @toshihikoyanase Thank you for the comments and sorry for the late reply. I just missed the former comment. 

I agree with providing the Sobol sampler as an integration module. Since it might take some time before I implement it, in case you need it soon, please feel free to implement it on your side. I am aiming to implement it by the end of this year.

**Comment by kstoneriv3** [comment]
I have a question on parallelized optimization. In parallel optimization settings, every Sobol sampler in each process must share the same random seed. What do you believe is the best way to share the random seed of the Sobol sampler?

**Comment by tupui** [comment]
FYI, there will be soon QMC in scipy with Sobol'. https://github.com/scipy/scipy/pull/10844
I would suggest you go with this instead of `sobol_seq` as the sequence misses the first point (zero). This is a bigger problem as it may sound as you can read in scipy's PR. Appart from scipy, this implementation is expected to be pushed to pytorch (this PR is based on original work in pytorch). But the version in scipy would allow more flexibility as more QMC samplers are available (LHS, Halton).

**Comment by kstoneriv3** [comment]
That is great news! Using scipy implementation sounds like a more reasonable option. Thank you for letting me know!

**Comment by Balandat** [comment]
The scipy PR has been ready for a long time, not quite sure what the holdup is to get this merged in. It could be helpful if people who would like to use it were to comment on the issue to give it some sense of urgency...

**Comment by rgommers** [comment]
The `scipy.stats.qmc` module is a thing now, PR is merged. It will be in the 1.7.0 release (which is still a while away, probably May) - but it'll be in the nightly wheels soon (https://anaconda.org/scipy-wheels-nightly/), probably tomorrow.

**Comment by tupui** [comment]
Let me know if you need help. I can take some time to propose a PR.

**Comment by kstoneriv3** [comment]
That is great news!! I will implement Sobol sampler for Optuna soon using `scipy.stats.qmc` soon!

**Comment by imbalu007** [comment]
> I implemented `SobolSampler` for optuna using a package [`sobol_seq`](https://github.com/naught101/sobol_seq), which is 1) lightweight and 2) distributed under MIT license. My implementation can be checked at [my repository](https://github.com/kstoneriv3/optuna/tree/sobol-sampler/optuna/samplers/_sobol).

@kstoneriv3 , would it be possible to share your code? I am getting a 'Page not found' for the link to your repository where you implemented sobol sampling

**Comment by kstoneriv3** [comment]
@imbalu007 I do not think it is possible in the current framework of optuna as it only supports the following 5 distributions. When you use `QMCSampler` you can only sample from these distributions.
- Categorical
- Uniform
- log-Uniform
- Integer (discretized uniform dist.)
- log-Integer (discretized log-uniform dist.)

**Comment by imbalu007** [comment]
@kstoneriv3 , is there any limitation to use SciPy's distributions (https://docs.scipy.org/doc/scipy/reference/stats.html#statistical-functions-scipy-stats) and then call ppf() to calculate inverse CDF with your sobol sample?

**Comment by tupui** [comment]
@imbalu007 you should ping members of the optuna team with your questions, or open a new issue since this one is closed and might not get traction.

For SciPy, you can use the inverse CDF or use [`scipy.stats.NumericalInverseHermite`](http://scipy.github.io/devdocs/reference/generated/scipy.stats.NumericalInverseHermite.html). SciPy 1.8 will come with more options from UNURAN.
